File: 983. Minimum Cost For Tickets/983. Minimum Cost For Tickets top-down.java
class Solution {
    private int[] days, costs, memo;
    private int n;

    private int dp(int day) {
        if (day > days[n - 1]) {
            return 0;
        }
        if (day == days[n - 1]) {
            return Math.min(Math.min(costs[0], costs[1]), costs[2]);
        }
        int i = 0;
        while (i < n && days[i] < day)
            i++;
        day = days[i];
        if (memo[day] == 0) {
            memo[day] = Math.min(Math.min(costs[0] +
                    dp(day + 1), costs[1] + dp(day + 7)), costs[2] + dp(day + 30));
        }
        return memo[day];
    }

    public int mincostTickets(int[] days, int[] costs) {
        this.days = days;
        this.costs = costs;
        this.n = days.length;
        this.memo = new int[366];
        return dp(days[0]);
    }
}

File: 983. Minimum Cost For Tickets/983. Minimum Cost For Tickets top-down.py
from functools import lru_cache
from typing import List


class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        @lru_cache(None)
        def dp(day: int) -> int:
            if day > days[-1]:
                return 0
            if day == days[-1]:
                return min(costs)
            else:
                i = 0
                while i < len(days) and days[i] < day:
                    i += 1
                if i == len(days):
                    return 0
                day = days[i]
            return min(costs[0]+dp(day+1), costs[1]+dp(day+7), costs[2]+dp(day+30))
        return dp(days[0])


s = Solution()
days = [1, 4, 6, 7, 8, 20]
costs = [2, 7, 15]
print(s.mincostTickets(days, costs))
